### PR TITLE
occlusions added

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

**Comment by BernhardEgger** [comment]
Dear Arneli,

thanks for taking time to improve on your pull request. Now it is much better visible what you changed and we can start to dicsuss on that.

There are still some things to improve before we can merge those new nice features.

I think it is not necessary to rename README.md to TODO.txt in folder bip

My main concern is that bfm and face12 are now always rendered together. This is specific to your usecase and might be unecessary for most other usecases. Please exclude all changes that always render both models. This makes it not possible to use other models anymore.

Please apply only the necessary changes to example config files. Changes in Landmarks and Models, number of ids, rendersettinges path of bip, yaw-distribution etc. are not necessary.

Please add a config file to generate a set of occlusion faces e.g. with the name "full_dataset_controlled".

Please remove the file "Release Notes" and add it in form of comments to the occlusion example config file - see other config files for comparison.

What is the source and license of the occlusion sample files?

There is a typo occolusion at various points.

Best
Bernhard

**Comment by BernhardEgger** [comment]
I just realized that this PR is removing the controlled setting by commenting it out. 
I'm fine if the occlusions are only supported in conjunction with the random setting (like for expressions), but we need to keep the controlled setting as is.

**Comment by ghost** [comment]
Hello Bernhard,
The code is now so that he uses only one model in the controlled setting and does not render an occlusion (according to your last post). In the config files, fewer landmarks must be used because the 'face12' does not recognize a few (eg right.ear.lobule.attachement). I have all the pictures of google. Unfortunately, I have not noted their source.

Best
Arneli

File: data/bip/README.txt [hunk 1]
```diff
@@ -0,0 +1 @@
+replace this file with '.rps' files of the Basel Illumination Prior!
```
Comment (BernhardEgger) [inline data/bip/README.txt:1]: This change is unecessary

File: data/config_files/example_config_controlled.json [hunk 2]
```diff
@@ -25,7 +25,7 @@
   "morphable-model-parameters": {
     "number-of-ids-to-create": 1,
     "number-of-samples-per-id": 1,
-    "model-filename": "data/bfm2017/model2017-1_bfm_nomouth.h5",
```
Comment (BernhardEgger) [inline data/config_files/example_config_controlled.json:28]: unecessary

File: data/config_files/example_config_random.json [hunk 3]
```diff
@@ -23,18 +23,18 @@
...
     "render": true,
-    "render-depth": true,
-    "render-color-correspondence-image": true,
-    "render-normals": true,
-    "render-albedo": true,
-    "render-illumination": true
+    "render-depth": false,
+    "render-color-correspondence-image": false,
+    "render-normals": false,
+    "render-albedo": false,
```
Comment (BernhardEgger) [inline data/config_files/example_config_random.json:29]: unecessary

File: data/config_files/example_config_random.json [hunk 4]
```diff
@@ -23,18 +23,18 @@
...
+    "render-illumination": false
   },
   "morphable-model-parameters": {
     "__comment-1" : "This block configures how face identities are generated and if they have expressions.",
     "__comment-2" : "You need to download the Basel Face Model and cite the paper in your publication.",
     "number-of-ids-to-create": 5,
     "number-of-samples-per-id": 10,
-    "model-filename": "data/bfm2017/model2017-1_bfm_nomouth.h5",
+    "model-filename": "data/bfm2017/model2017-1_face12_nomouth.h5",
```
Comment (BernhardEgger) [inline data/config_files/example_config_random.json:37]: unecessary

File: data/config_files/example_config_random.json [hunk 5]
```diff
@@ -47,7 +47,7 @@
     "__comment-3" : "no-color removes replaces all illumination components by the mean over all color channels (non-colored illumination)",
     "__comment-4" : "if fix-energy is set the value of the energy of Spherical Harmonics illumination is set to the value defined by energy-value",
     "illumination-type": "empirical",
-    "illumination-prior-directory": "data/bip/parameters/",
+    "illumination-prior-directory": "data/bip/parameters",
```
Comment (BernhardEgger) [inline data/config_files/example_config_random.json:50]: unecessary

File: data/config_files/full_dataset_controlled.json [hunk 6]
```diff
@@ -18,7 +18,7 @@
   "morphable-model-parameters": {
     "number-of-ids-to-create": 1,
     "number-of-samples-per-id": 1,
-    "model-filename": "data/bfm2017/model2017-1_bfm_nomouth.h5",
+    "model-filename": "data/bfm2017/model2017-1_face12_nomouth.h5",
```
Comment (BernhardEgger) [inline data/config_files/full_dataset_controlled.json:21]: unecessary

File: data/config_files/full_dataset_random.json [hunk 7]
```diff
@@ -0,0 +1,121 @@
+{
```
Comment (BernhardEgger) [inline data/config_files/full_dataset_random.json:1]: rename it to example_config_random_occlusion
the full dataset term is used in conjunction with the paper

File: src/main/scala/faces/apps/ControlledFaces.scala [hunk 8]
```diff
@@ -84,30 +85,34 @@ object ControlledFaces extends App {
                 }
                 else if(faceCenter == "landmark")
                 {
-                  helpers.centerLandmark(uncentered)
+                  helpers.centerLandmark_face12(uncentered)
```
Comment (BernhardEgger) [inline src/main/scala/faces/apps/ControlledFaces.scala:88]: unecessary

File: src/main/scala/faces/apps/ControlledFaces.scala [hunk 9]
```diff
@@ -84,30 +85,34 @@ object ControlledFaces extends App {
...
                 }
                 else{
                   uncentered
                 }
 
                 val imageData =
-                  for((postfix, currentRenderer) <- helpers.renderingMethods) yield {
+                  for((postfix, currentRenderer) <- helpers.face12_renderingMethods) yield {
```
Comment (BernhardEgger) [inline src/main/scala/faces/apps/ControlledFaces.scala:95]: unecessary

File: src/main/scala/faces/apps/ControlledFaces.scala [hunk 10]
```diff
@@ -84,30 +85,34 @@ object ControlledFaces extends App {
...
+                        RGBA.Black
+                      })
+                      (rendering, postfix, rendering.zip(dummy).map(p => if (p._1.a < 0.5) RGBA.Black else RGBA.White))
                     }
                   }
 
                 // write images and their parameters
                 println(s"Generating \t ID:$id \t Sample:$n")
-                for((img, postifx) <- imageData) {
-                  helpers.writeRenderParametersAndLandmarks(centered, id, n)
```
Comment (BernhardEgger) [inline src/main/scala/faces/apps/ControlledFaces.scala:109]: remove_face12

File: src/main/scala/faces/apps/ControlledFaces.scala [hunk 11]
```diff
@@ -84,30 +85,34 @@ object ControlledFaces extends App {
...
 
                 val imageData =
-                  for((postfix, currentRenderer) <- helpers.renderingMethods) yield {
+                  for((postfix, currentRenderer) <- helpers.face12_renderingMethods) yield {
                     if (bg && postfix == "") {
                       require(helpers.loadBgs.nonEmpty, "no Background files with type " + cfg.backgrounds.bgType + " found in " + cfg.backgrounds.bgPath)
                       val BG = helpers.loadBgs(b)
                       val controlledBGimg = PixelImageIO.read[RGBA](BG).get.resample(imageWidth, imageHeight)
-                      (currentRenderer.renderImage(centered).zip(controlledBGimg).map(p => if (p._1.a < 0.5) p._2 else p._1), postfix)
+                      (currentRenderer.renderImage(centered).zip(controlledBGimg).map(p => if (p._1.a < 0.5) p._2 else p._1), postfix, currentRenderer.renderImage(centered).zip(controlledBGimg).map(p => if (p._1.a < 0.5) RGBA.Black else RGBA.White))
```
Comment (BernhardEgger) [inline src/main/scala/faces/apps/ControlledFaces.scala:100]: use a variable to avoid long line

File: src/main/scala/faces/apps/ControlledFaces.scala [hunk 12]
```diff
@@ -84,30 +85,34 @@ object ControlledFaces extends App {
...
                     }
                     else {
-                      (currentRenderer.renderImage(centered), postfix)
+                      val rendering = currentRenderer.renderImage(centered)
+                      val dummy = PixelImage(rendering.width, rendering.height, (x, y) => {
```
Comment (BernhardEgger) [inline src/main/scala/faces/apps/ControlledFaces.scala:104]: unecessary

File: src/main/scala/faces/apps/RandomFaces.scala [hunk 13]
```diff
@@ -80,12 +80,12 @@ object RandomFaces extends App {
         // put RenderParameters together of all its randomized parts
         val uncentered = RenderParameter(rndPose, view, rndCamera, rndIll, directionalLight, momoInstance, ImageSize(imageWidth, imageHeight), colorTransform)
 
-        // move face in the middle of the image
+
```
Comment (BernhardEgger) [inline src/main/scala/faces/apps/RandomFaces.scala:83]: unecessary - removed comment

File: src/main/scala/faces/apps/RandomFaces.scala [hunk 14]
```diff
@@ -80,12 +80,12 @@ object RandomFaces extends App {
...
         val centered = if (faceCenter == "facebox") {
           helpers.centerFaceBox(uncentered)
         }
         else if (faceCenter == "landmark") {
-          helpers.centerLandmark(uncentered)
```
Comment (BernhardEgger) [inline src/main/scala/faces/apps/RandomFaces.scala:88]: unecessary

File: src/main/scala/faces/apps/RandomFaces.scala [hunk 15]
```diff
@@ -99,24 +99,38 @@ object RandomFaces extends App {
...
               require(helpers.loadBgs.nonEmpty, "no Background files with type " + cfg.backgrounds.bgType + " found in " + cfg.backgrounds.bgPath)
               val rndBG = helpers.loadBgs(rnd.scalaRandom.nextInt(helpers.loadBgs.length))
               val rndBGimg = PixelImageIO.read[RGBA](rndBG).get.resample(imageWidth, imageHeight)
-              (currentRenderer.renderImage(rps).zip(rndBGimg).map(p => if (p._1.a < 0.5) p._2 else p._1), postfix)
+              val img = currentRenderer.renderImage(rps).zip(rndBGimg).map(p => if (p._1.a < 0.5) p._2 else p._1)
+              val mask = currentRenderer.renderImage(centered).zip(rndBGimg).map(p => if (p._1.a < 0.5) RGBA.Black else RGBA.White)
+              helpers.writeRenderParametersAndLandmarks_face12(rps, id, n, mask, "initial")
+              val mask_with_occlusions_and_path = helpers.writeImg_face12(img, id, n, postfix, cfg.landmarkTags, mask)
+              helpers.writeRenderParametersAndLandmarks_face12(rps, id, n, mask_with_occlusions_and_path._1)
+              // here, we switch to the bfm
```
Comment (BernhardEgger) [inline src/main/scala/faces/apps/RandomFaces.scala:112]: remove - one model at a time

File: src/main/scala/faces/apps/RandomFaces.scala [hunk 16]
```diff
@@ -99,24 +99,38 @@ object RandomFaces extends App {
...
-              (currentRenderer.renderImage(rps), postfix)
+              (currentRenderer.renderImage(rps), postfix, currentRenderer.renderImage(rps))
             }
           }
 
-        // write images and their parameters
-        println(s"Generating \t ID:$id \t Sample:$n")
-        for ((img, postifx) <- imageData) {
-          helpers.writeRenderParametersAndLandmarks(rps, id, n)
-          helpers.writeImg(img, id, n, postifx)
```
Comment (BernhardEgger) [inline src/main/scala/faces/apps/RandomFaces.scala:118]: why is this removed?

File: src/main/scala/faces/apps/RandomFaces.scala [hunk 17]
```diff
@@ -99,24 +99,38 @@ object RandomFaces extends App {
         ))
 
         val imageData =
```
Comment (BernhardEgger) [inline src/main/scala/faces/apps/RandomFaces.scala:101]: I would prefer to integrate the occlusion as done for the expressions.
Perhaps use a seperate renderer as for other modalities and see occlusion as a modality?

File: src/main/scala/faces/settings/FacesSettings.scala [hunk 18]
```diff
@@ -27,17 +27,31 @@ abstract class FacesSettings{
...
 
 case class OutputLocation(
                            outPath: String
                          ) {
-  def outImgPath: String = outPath + "img/"
-  def outRpsPath: String = outPath + "rps/"
-  def outCSVPath: String = outPath + "csv/"
-  def outTLMSPath: String = outPath + "tlms/"
-
+  def outImgPath_bfm: String = outPath + "bfm/img_original/"
```
Comment (BernhardEgger) [inline src/main/scala/faces/settings/FacesSettings.scala:37]: unecessary no bfm and face_12 - please keep paths

**Review by BernhardEgger — CHANGES_REQUESTED** [review]
Hi Arneli,
there are still plenty of things that change a lot and are not connected to the new functionality. It seems that you tuned it to a very specific use case.
I would like to keep everything as is, and just add the functionality to render occlusions.
Sorry but it seems that this needs still a lot of work.
I discussed with Adam and we agreed that we would like to have the minimal set of changes to include the new functionality. Please see Adam - he will help you with this.
I marked in the code which parts I think are not needed.
Perhaps it also makes sense to rethink how the occlusions are put on top of the image - currently, this kills some previous functionality and is hard to extend
Best
Bernhard

**Comment by ghost** [comment]
Hello Bernhard
I talked to Adam and I completely rewrote the software because I misunderstood the requirements. Is [this](https://github.com/Arneli/pull_request) what you imagined? If so, I will improve the software a little bit and start a new PR.

PS: I work with Windows and git automatically renames .md files to .txt files. (LF will be replaced by CRLF in git) that is the warning by git:
`warning: LF will be replaced by CRLF in data/bip/README.txt.
The file will have its original line endings in your working directory.`

**Comment by BernhardEgger** [comment]
Hi Arneli,
thanks for investing time to do that. It is however hard for me to tell from the other repository if that is what I thought of...
On the first gist it looks like some of my comments where not incorporated (like not changing config files and adding a new config file). I also don't get imidiatly where the occlusion now comes in.
Just update that PR or make a new one - then it is much easier for us to judge on it and put comments...
Best
Bernhard
